Source/Core/Common/VR/DolphinVR.cpp

```diff
@@ -0,0 +1,219 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <cstring>
+
+#include "Common/VR/DolphinVR.h"
+#include "Common/VR/VRBase.h"
+#include "Common/VR/VRInput.h"
+#include "Common/VR/VRMath.h"
+#include "Common/VR/VRRenderer.h"
+
+namespace Common::VR
+{
+Base* s_module_base = NULL;
+Input* s_module_input = NULL;
+Renderer* s_module_renderer = NULL;
+bool s_platform_flags[PLATFORM_MAX];
+bool s_enabled = false;
+
+void OXR_CheckErrors(XrResult result, const char* function)
+{
+  if (XR_FAILED(result))
+  {
+    char errorBuffer[XR_MAX_RESULT_STRING_SIZE];
+    xrResultToString(s_module_base->GetEngine()->app_state.instance, result, errorBuffer);
+    ERROR_LOG_FMT(VR, "error: {} {}", function, errorBuffer);
+  }
+}
+
+static void (*UpdateInput)(int id, int l, int r, float x, float y, float jlx, float jly, float jrx,
+                           float jry);
+
+/*
+================================================================================
+
+VR app flow integration
+
+================================================================================
+*/
+
+bool IsEnabled()
+{
+#ifdef OPENXR
+  return s_enabled;
+#else
+  return false;
+#endif
+}
+
+#ifdef ANDROID
+void InitOnAndroid(JNIEnv* env, jobject obj, const char* vendor, int version, const char* name)
+{
+  // Do not allow second initialization
+  if (s_platform_flags[PLATFORM_STATUS_INITIALIZED])
+  {
+    s_module_renderer->SetConfigInt(ConfigInt::CONFIG_VIEWPORT_VALID, false);
+    return;
+  }
+
+  // Set platform flags
+  if ((strcmp(vendor, "Meta") == 0) || (strcmp(vendor, "Oculus") == 0))
+  {
+    s_platform_flags[PLATFORM_CONTROLLER_QUEST] = true;
+    s_platform_flags[PLATFORM_EXTENSION_PERFORMANCE] = true;
+  }
+
+  // Allocate modules
+  s_module_base = new Base();
+  s_module_input = new Input();
+  s_module_renderer = new Renderer();
+  s_platform_flags[PLATFORM_STATUS_INITIALIZED] = true;
+  s_module_renderer->SetConfigFloat(CONFIG_CANVAS_DISTANCE, 4.0f);
```

Comment:
> If you are talking about Common/VR then it is a common topic with OpenXR. The OpenXR is connection of render/input/tracking, splitting it into different modules usually leads after some revisions into input delays and lags. For great experience it is better to keep the code in the same module.

Your code is already split into multiple modules, what @iwubcode is saying is that these modules probably shouldn't all be included under Core/Common/VR. I'll review the modules and see if there is indeed a need to move them.

Reply:
@iwubcode, could you point me where would be the correct places for my 6 classes?  (See https://github.com/dolphin-emu/dolphin/pull/12564#issuecomment-1946031204 for description what the classes do).

I somehow do not understand the Dolphin SW architecture.

Reply:
Let's wait a bit with that. One of the reasons @iwubcode is because it contains rendering code, but I want to see if we can get rid of that. Maybe we won't have to move stuff in the end, we'll see once all the important stuff has been taken care of.